Storage/SQL/SQLiteLogins.swift

```diff
@@ -0,0 +1,181 @@
+        let sql = "SELECT username, password, hostname, httpRealm, formSubmitUrl, usernameField, passwordField FROM \(table.name) WHERE hostname = ? ORDER BY timeLastUsed DESC"
+        let args: [AnyObject?] = [protectionSpace.host]
+        return db.runQuery(sql, args: args, factory: SQLiteLogins.LoginDataFactory)
+    }
+
+    public func getUsageDataForLogin(login: LoginData) -> Deferred<Result<LoginUsageData>> {
+        let sql = "SELECT * FROM \(table.name) WHERE hostname = ? AND username IS ? LIMIT 1"
+        let args: [AnyObject?] = [login.hostname, login.username]
+        return db.runQuery(sql, args: args, factory: SQLiteLogins.LoginUsageDataFactory) >>== { value in
+            return deferResult(value[0]!)
+        }
+    }
+
+    public func addLogin(login: LoginData) -> Success {
+        var args = [AnyObject?]()
+        args.append(login.hostname)
+        args.append(login.httpRealm)
+        args.append(login.formSubmitUrl)
+        args.append(login.usernameField)
+        args.append(login.passwordField)
+
+        if var login = login as? SyncableLoginData {
+            if login.guid == nil {
+                login.guid = NSUUID().UUIDString
+            }
+            args.append(login.guid)
+        } else {
+            args.append(NSUUID().UUIDString)
+        }
+
+        let date = NSNumber(unsignedLongLong: NSDate.nowMicroseconds())
+        args.append(date) // timeCreated
+        args.append(date) // timeLastUsed
+        args.append(date) // timePasswordChanged
+        args.append(login.username)
+        args.append(login.password)
+
+        return db.run("INSERT INTO \(table.name) (hostname, httpRealm, formSubmitUrl, usernameField, passwordField, guid, timeCreated, timeLastUsed, timePasswordChanged, username, password) VALUES (?,?,?,?,?,?,?,?,?,?,?)", withArgs: args)
+    }
+
+    public func addVisitFor(login: LoginData) -> Success {
+        let date = NSNumber(unsignedLongLong: NSDate.nowMicroseconds())
+        return db.run("UPDATE \(table.name) SET timeLastUsed = ? WHERE hostname = ? AND username IS ?", withArgs: [date, login.hostname, login.username])
+    }
+
+    public func updateLogin(login: LoginData) -> Success {
+        let date = NSNumber(unsignedLongLong: NSDate.nowMicroseconds())
+        var args = [AnyObject?]()
+        args.append(login.httpRealm)
+        args.append(login.formSubmitUrl)
+        args.append(login.usernameField)
+        args.append(login.passwordField)
+        args.append(date) // timePasswordChanged
+        args.append(login.password)
+        args.append(login.hostname)
+        args.append(login.username)
+
+        return db.run("UPDATE \(table.name) SET httpRealm = ?, formSubmitUrl = ?, usernameField = ?, passwordField = ?, timePasswordChanged = ?, password = ? WHERE hostname = ? AND username IS ?", withArgs: args)
+    }
+
+    public func removeLogin(login: LoginData) -> Success {
+        var args = [AnyObject?]()
```

Comment:
We have a typealias of `Args` that you can use throughout instead of `[AnyObject?]`.